lib/linked_list.rb

```diff
@@ -7,39 +7,76 @@ def initialize
     @head = nil
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity O(1)
+  # Space complexity O(1)
   def add_first(data)
-
+    @head = Node.new(data, @head)
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(1)
+  # Space complexity - O(1)
   def get_first
-
+    if !(@head.nil?)
+      return @head.data
+    end
+    @head
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(n)
+  # Space complexity - O(1)
   def length
-    return 0
+    temp = @head
+    count = 0
+
+    while temp 
+      count += 1
+      temp = temp.next
+    end
+
+    return count
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(n)
+  # Space complexity - O(1)
   def add_last(data)
+    new_node = Node.new(data, @head)
```

Comment:
The new node will point the the head as the next?  This will make a circular list.